cranelift-codegen/src/isa/x86/abi.rs

```diff
@@ -547,43 +643,115 @@ fn insert_common_epilogues(
+                    isa,
+                    &mut cfa_state.clone(),
+                );
+
+                if let Some(following_inst) = following_inst {
+                    if let Some(ref mut frame_layout) = pos.func.frame_layout {
+                        frame_layout.instructions.insert(
+                            following_inst,
+                            vec![FrameLayoutChange::Restore].into_boxed_slice(),
+                        );
+                    }
+                }
             }
         }
     }
 }
 
 /// Insert an epilogue given a specific `return` instruction.
 /// This is used by common calling conventions such as System V.
+/// TODO implement and handle _cfa_state more than one epilogue.
 fn insert_common_epilogue(
     inst: ir::Inst,
     stack_size: i64,
     pos: &mut EncCursor,
     reg_type: ir::types::Type,
     csrs: &RegisterSet,
+    isa: &dyn TargetIsa,
+    cfa_state: &mut CFAState,
 ) {
+    let word_size = isa.pointer_bytes() as isize;
     if stack_size > 0 {
         pos.ins().adjust_sp_up_imm(Imm64::new(stack_size));
     }
 
     // Pop all the callee-saved registers, stepping backward each time to
     // preserve the correct order.
     let fp_ret = pos.ins().x86_pop(reg_type);
+    // account for CFA state in the reverse of `insert_common_prologue`.
+    cfa_state.current_depth += word_size;
+    cfa_state.cf_ptr_offset -= word_size;
+    // and now that we're going to overwrite `rbp`, `rsp` is the only way to get to the call frame.
+    cfa_state.cf_ptr_reg = RU::rsp as RegUnit;
+
     pos.prev_inst();
 
     pos.func.locations[fp_ret] = ir::ValueLoc::Reg(RU::rbp as RegUnit);
     pos.func.dfg.append_inst_arg(inst, fp_ret);
 
     for reg in csrs.iter(GPR) {
         let csr_ret = pos.ins().x86_pop(reg_type);
+        cfa_state.current_depth += word_size;
         pos.prev_inst();
 
         pos.func.locations[csr_ret] = ir::ValueLoc::Reg(reg);
         pos.func.dfg.append_inst_arg(inst, csr_ret);
     }
+
+    if let Some(ref mut frame_layout) = pos.func.frame_layout {
+        // Validity checks - if we accounted correctly, CFA state at a return will match CFA state
+        // at the entry of a function.
+        //
+        // current_depth starts assuming a return address is pushed, and cf_ptr_offset is one
+        // pointer below current_depth.
+        assert_eq!(cfa_state.current_depth, -word_size);
+        assert_eq!(cfa_state.cf_ptr_offset, word_size);
+
+        let new_cfa = FrameLayoutChange::CallFrameAddressAt {
+            reg: cfa_state.cf_ptr_reg,
+            offset: cfa_state.cf_ptr_offset,
+        };
+
+        frame_layout
+            .instructions
+            .entry(inst)
+            .and_modify(|insts| {
+                let mut new_instructions = insts.to_vec();
+                new_instructions.push(new_cfa);
+                *insts = new_instructions.into_boxed_slice();
+            })
+            .or_insert_with(|| vec![new_cfa].into_boxed_slice());
```

Comment:
```suggestion
            .or_insert_with(|| Box::new([new_cfa]) as Box<[_]>);
```

Before the `vec!` macro created a boxed slice, turned it into a vec and then you turned it into a boxed slice again. This change should not matter for perf in release mode, but could save a bit of time during compilation for optimizing it.